File: wasm-runner/src/field.ts
/**
 * BN254 Field Element Utilities
 */

export const MODULUS = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;
const R = 6350874878119819312338956282401532410528162663560392320966563075034087161851n;
const R_INV = 9915499612839321149637521777990102151350674507940716049588462388200839649614n;
const MASK_64 = (1n << 64n) - 1n;

export type FieldElement = [bigint, bigint, bigint, bigint];

export function toMontgomery(value: bigint): FieldElement {
  let v = value % MODULUS;
  if (v < 0n) v += MODULUS;
  const mont = (v * R) % MODULUS;
  return [
    mont & MASK_64,
    (mont >> 64n) & MASK_64,
    (mont >> 128n) & MASK_64,
    (mont >> 192n) & MASK_64,
  ];
}

export function fromMontgomery(limbs: FieldElement): bigint {
  const mont = limbs[0] | (limbs[1] << 64n) | (limbs[2] << 128n) | (limbs[3] << 192n);
  return (mont * R_INV) % MODULUS;
}

export function fieldToHex(limbs: FieldElement): string {
  return '0x' + fromMontgomery(limbs).toString(16).padStart(64, '0');
}

export function parseFieldElement(str: string): FieldElement {
  return toMontgomery(BigInt(str));
}

File: native-runner/src/field.rs
//! BN254 Field Element Utilities
//!
//! Handles conversion between decimal strings and Montgomery form.

use ark_bn254::Fr;
use ark_ff::PrimeField;

/// Field element as 4 x u64 limbs in Montgomery form
#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct FieldElement {
    pub limbs: [u64; 4],
}

impl FieldElement {
    /// Create a zero field element
    pub fn zero() -> Self {
        Self { limbs: [0; 4] }
    }

    /// Convert to hex string for JSON output
    pub fn to_hex(&self) -> String {
        format!(
            "0x{:016x}{:016x}{:016x}{:016x}",
            self.limbs[3], self.limbs[2], self.limbs[1], self.limbs[0]
        )
    }
}

/// Convert a decimal string to Montgomery form field element
pub fn to_montgomery(value: &str) -> FieldElement {
    // Parse the decimal string
    let fr = if value.starts_with("0x") || value.starts_with("0X") {
        // Hex input
        let hex = &value[2..];
        let bytes = hex::decode(hex).expect("Invalid hex string");
        Fr::from_be_bytes_mod_order(&bytes)
    } else {
        // Decimal input
        let n: num_bigint::BigUint = value.parse().expect("Invalid decimal string");
        let bytes = n.to_bytes_be();
        Fr::from_be_bytes_mod_order(&bytes)
    };

    // Extract Montgomery form limbs
    let limbs = fr.0.0;
    FieldElement { limbs }
}

/// Convert Montgomery form field element back to decimal string
pub fn from_montgomery(fe: &FieldElement) -> String {
    use ark_ff::BigInteger;

    let fr = Fr::new_unchecked(ark_ff::BigInt::new(fe.limbs));
    // Convert to standard form and then to string
    let big_int = fr.into_bigint();

    // Convert limbs to big integer
    let mut result = num_bigint::BigUint::from(0u64);
    for (i, limb) in big_int.0.iter().enumerate() {
        result += num_bigint::BigUint::from(*limb) << (64 * i);
    }

    result.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_roundtrip() {
        let values = ["0", "1", "2", "42", "12345678901234567890"];
        for v in values {
            let fe = to_montgomery(v);
            let back = from_montgomery(&fe);
            assert_eq!(v, back, "Roundtrip failed for {}", v);
        }
    }

    #[test]
    fn test_known_values() {
        // 2 in Montgomery form
        let two = to_montgomery("2");
        assert_eq!(two.limbs[0], 0x592c68389ffffff6);

        // 3 in Montgomery form
        let three = to_montgomery("3");
        assert_eq!(three.limbs[0], 0x05c29c54effffff1);
    }
}
